static/app/components/searchQueryBuilder/useQueryBuilderState.tsx

```diff
@@ -0,0 +1,56 @@
+import {type Reducer, useCallback, useMemo, useReducer} from 'react';
+
+import type {
+  ParseResultToken,
+  Token,
+  TokenResult,
+} from 'sentry/components/searchSyntax/parser';
+
+type QueryBuilderState = {
+  focus: null; // TODO(malwilley): Implement focus state
+  query: string;
+};
+
+type DeleteTokenAction = {
+  token: ParseResultToken;
+  type: 'DELETE_TOKEN';
+};
+
+export type QueryBuilderActions = DeleteTokenAction;
+
+function removeQueryToken(query: string, token: TokenResult<Token>): string {
+  return (
+    query.substring(0, token.location.start.offset) +
+    query.substring(token.location.end.offset)
+  );
+}
+
+export function useQueryBuilderState({initialQuery}: {initialQuery: string}) {
+  const initialState: QueryBuilderState = {query: initialQuery, focus: null};
+
+  const reducer: Reducer<QueryBuilderState, QueryBuilderActions> = useCallback(
+    (state, action): QueryBuilderState => {
+      switch (action.type) {
+        case 'DELETE_TOKEN':
+          return {
+            ...state,
+            query: removeQueryToken(state.query, action.token),
+            focus: null,
+          };
+        default:
+          return state;
+      }
+    },
+    []
+  );
+
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  return useMemo(
+    () => ({
+      state,
+      dispatch,
+    }),
+    [state, dispatch]
+  );
```

Comment:
does it really matter if the return object is stable if you're just going to restructure it?

Reply:
Nope you're right, not necessary. Removed